enumerable_methods.rb

```diff
@@ -0,0 +1,145 @@
+module Enumerable
+  # my each
+  def my_each
+    return to_enum unless block_given?
+    size.times do |n|
+      yield to_a[n]
+    end
+    self
+  end
+
+  # my_each_with_index
+  def my_each_with_index
+    return to_enum unless block_given?
+    n = 0
+    while n < to_a.length
+      yield(to_a[n], n)
+      n += 1
+    end
+    self
+  end
+
+  # my_select
+  def my_select
+    return to_enum unless block_given?
+    selected = is_a?(Array || Range) ? [] : {}
+    if is_a?(Array)
+      my_each do |i|
+        selected.push(i) if yield(i)
+      end
+    else
+      my_each do |key, val|
+        selected[key] = val if yield(key, val)
+      end
+    end
+    selected
+  end
+
+  # my_all?
+  def my_all?(pattern = nil)
+    result = true
+    my_each do |i|
+      block_given? && pattern.nil? ? result = yield(i) : next 
+      case pattern
+      when nil
+          result = false if !i
+          
+        when Regexp
+        result = false unless i.match(pattern)
+      when Class
+        result = false unless i.is_a?(pattern)
+      when String, Numeric
+        result = false unless i == pattern
+      when nil 
+        result = false unless yield(i)
+      
+      end
+      break if result != true
+    end
+    result
+  end
+
+  # my_any?
+  def my_any?(pattern = nil)
+    result = false
+    my_each do |i|
+      case pattern
+      when nil
+        result = true if i
+      when Regexp
+        result = true if i.match(pattern)
+      when Class
+        result = true if i.is_a?(pattern)
+      when String, Numeric
+        result = true if i == pattern
+      end
+      result = yield(i) if block_given?
+      break if result
+    end
+    result
+  end
+
+  # my_none?
+  def my_none?(pattern = nil)
+    result = true
+    my_each do |i|
+      case pattern
+      when nil
+        result = false if i
+      when Regexp
+        result = false if i.match(pattern)
+      when Class
+        result = false if i.is_a?(pattern)
+      when String, Numeric
+        result = false if i == pattern
+      end
+    result = !yield(i) if block_given?
+      break if !result
+    end
+    result
+  end
+
+  # my_count
+  def my_count(check = nil)
+    counter = 0
+    my_each do |i|
+      counter += 1 if block_given? && yield(i)
+      counter += 1 if check == i
+      counter = size if check.nil? && !block_given?
+    end
+    counter
+  end
+
+  # my_map
+  def my_map
+    return to_enum unless block_given?
+    result = []
+    my_each do |i|
+      block_given? ? result.push(yield(i)) : to_enum
+    end
+    result
+  end
+
+  # Proc for my_map as requested in the project assignment
+  proc { |i| i**3 }
+
+  # my_inject
+  def my_inject(sym = nil, start = nil)
+    acc = start ? start : 0 
+    (to_a.length.times).my_each do |i|
+      acc = yield(acc, to_a[i]) if block_given?
+      acc =  acc.send(sym, to_a[i]) if sym && start.nil?
+      acc = acc.send(sym, to_a[i]) if sym && start
+      end
+    acc
+  end
```

Comment:
#### my_inject 

- [ ] combines all elements of enum by applying a binary operation, specified by a block

     Given
       `  operation = proc { |sum, n| sum + n } `

     Expected result

      > (1..5).my_inject(4) { |prod, n| prod * n } should be equal (1..5).inject(4) { |prod, n| prod * n } => 480

- [ ] When a symbol is provided as an argument, my_inject should combine the elements of the enum using that symbol as a named method.

     > this [2, 5, 6].my_inject(:+) should be equal to this [2, 5, 6].inject(:+) => 13